File: Online_Judge_Problems/boj/boj_1931_회의실_배정/boj_1931.py
'''----------------------------------------------------
Sub  : [BOJ] 회의실 배정
Link : https://www.acmicpc.net/problem/1939
Level: 골드 5
Tag  : Python, greddy
Memo
----------------------------------------------------'''

import sys

TEST_MODE = True

def get_inputs():
    if TEST_MODE:
        with open('input.txt', 'r') as f:
            for line in f:
                yield line.rstrip('\n')
    else:
        for line in sys.stdin:
            yield line.rstrip('\n')

input_gen = get_inputs()

def input():
    return next(input_gen)


# input
n = int(input())
meetings = []

for _ in range(n):
    start, end = map(int, input().split())
    meetings.append((start, end))

# 정렬 기준을 함수로 정의
def sort_key(meeting):
    return (meeting[1], meeting[0])  # (끝나는 시간, 시작 시간)

# 정렬 수행
meetings.sort(key=sort_key)

end_time = 0
answer = 0

for start, end in meetings:
    if start >= end_time:
        end_time = end
        answer += 1

print(answer)

File: Online_Judge_Problems/boj/boj_1931_회의실_배정/boj_1931.js
/*-----------------------------------------------------
Sub  : [BOJ] 회의실 배정
Link : https://www.acmicpc.net/problem/1931
Level: 골드 5
Tag  : JS, greddy
Memo
-----------------------------------------------------*/

const TEST_MODE = true;

let input;

if (TEST_MODE) {
  const filePath = require('path').join(__dirname, 'input.txt');
  input = require('fs').readFileSync(filePath, 'utf-8').trim().split('\n');
}
else {
  input = require('fs').readFileSync(0, 'utf-8').trim().split('\n');
}


//input
const N = Number(input[0]);
const roomTimes = input.slice(1).map(line => line.trim().split(' ').map(Number));


// processing
roomTimes.sort((a, b) => {
  if (a[1] === b[1]) return a[0] - b[0];  // 종료 시간 같으면 시작 시간 오름차순
  return a[1] - b[1];
});

let answer = 0;
let currentEnd = 0;

for (let i = 0; i < N; i++) {
  const [start, end] = roomTimes[i];
  if (start >= currentEnd) {
    answer++;
    currentEnd = end;
  }
}

console.log(answer);
